Task 2 

Run the same code with a different loss function: 
Logistic loss as described in Brandon Amos blog and compare the results with above Task 1. 

You may need to modify the network architectures slightly with logit loss. 

Run the code for 5, 10 and 50 epochs and observe the results in both cases. 

How/why is the output different for both
cases? Try to find a suitable reason for both.


In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import numpy as np
import tqdm
import os
import wandb


# Hyperparameters
mb_size = 64
Z_dim = 1000
h_dim = 128
lr = 1e-3

# Load MNIST data
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Lambda(lambda x: x.view(-1))  # Flatten the 28x28 image to 784
])

train_dataset = datasets.MNIST(root='../MNIST', train=True, transform=transform, download=True)
train_loader = DataLoader(train_dataset, batch_size=mb_size, shuffle=True)

X_dim = 784  # 28 x 28

# Xavier Initialization
def xavier_init(m):
    if isinstance(m, nn.Linear):
        nn.init.xavier_normal_(m.weight)
        if m.bias is not None:
            nn.init.zeros_(m.bias)

# Generator
class Generator(nn.Module):
    def __init__(self, z_dim, h_dim, x_dim):
        super(Generator, self).__init__()
        self.fc1 = nn.Linear(z_dim, h_dim)
        self.fc2 = nn.Linear(h_dim, x_dim)
        self.apply(xavier_init)

    def forward(self, z):
        h = F.relu(self.fc1(z))
        out = torch.sigmoid(self.fc2(h))
        return out

# Discriminator
class Discriminator(nn.Module):
    def __init__(self, x_dim, h_dim):
        super(Discriminator, self).__init__()
        self.fc1 = nn.Linear(x_dim, h_dim)
        self.fc2 = nn.Linear(h_dim, 1)
        self.apply(xavier_init)

    def forward(self, x):
        h = F.relu(self.fc1(x))
        out = self.fc2(h)
        return out



# Training
def logisticGANTraining(G, D, loss_fn, train_loader):
    G.train()
    D.train()

    D_loss_real_total = 0
    D_loss_fake_total = 0
    G_loss_total = 0
    t = tqdm.tqdm(train_loader)
    
    for it, (X_real, labels) in enumerate(t):
        # Prepare real data
        X_real = X_real.float().to(device)

        # Sample noise and labels
        z = torch.randn(X_real.size(0), Z_dim).to(device)
        ones_label = torch.ones(X_real.size(0), 1).to(device)
        zeros_label = torch.zeros(X_real.size(0), 1).to(device)

        # ================= Train Discriminator =================
        G_sample = G(z)
        D_real = D(X_real)
        D_fake = D(G_sample.detach())

        D_loss_real = F.softplus(-D_real).mean()  # log(1 + e^{-D(x_real)})
        D_loss_fake = F.softplus(D_fake).mean()   # log(1 + e^{D(x_fake)})
        D_loss = D_loss_real + D_loss_fake
        D_loss_real_total += D_loss_real.item()
        D_loss_fake_total += D_loss_fake.item()

        D_solver.zero_grad()
        D_loss.backward()
        D_solver.step()

        # ================= Train Generator ====================
        z = torch.randn(X_real.size(0), Z_dim).to(device)
        G_sample = G(z)
        D_fake = D(G_sample)

        G_loss = F.softplus(-D_fake).mean()  # log(1 + e^{-D(G(z))})
        G_loss_total += G_loss.item()

        G_solver.zero_grad()
        G_loss.backward()
        G_solver.step()

    # ================= Logging =================
    D_loss_real_avg = D_loss_real_total / len(train_loader)
    D_loss_fake_avg = D_loss_fake_total / len(train_loader)
    D_loss_avg = D_loss_real_avg + D_loss_fake_avg
    G_loss_avg = G_loss_total / len(train_loader)

    wandb.log({
        "D_loss_real": D_loss_real_avg,
        "D_loss_fake": D_loss_fake_avg,
        "D_loss": D_loss_avg,
        "G_loss": G_loss_avg
    })

    return G, D, G_loss_avg, D_loss_avg
    


def save_sample(G, epoch, mb_size, Z_dim):
    out_dir = "out_vanila_GAN2"
    G.eval()
    with torch.no_grad():
        z = torch.randn(mb_size, Z_dim).to(device)
        samples = G(z).detach().cpu().numpy()[:16]

    fig = plt.figure(figsize=(4, 4))
    gs = gridspec.GridSpec(4, 4)
    gs.update(wspace=0.05, hspace=0.05)

    for i, sample in enumerate(samples):
        ax = plt.subplot(gs[i])
        plt.axis('off')
        ax.set_xticklabels([])
        ax.set_yticklabels([])
        ax.set_aspect('equal')
        plt.imshow(sample.reshape(28, 28), cmap='Greys_r')

    if not os.path.exists(f'{out_dir}'):
        os.makedirs(f'{out_dir}')

    plt.savefig(f'{out_dir}/{str(epoch).zfill(3)}.png', bbox_inches='tight')
    plt.close(fig)



########################### Main #######################################
wandb_log = True
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

# Instantiate models
G = Generator(Z_dim, h_dim, X_dim).to(device)
D = Discriminator(X_dim, h_dim).to(device)

# Optimizers
G_solver = optim.Adam(G.parameters(), lr=lr)
D_solver = optim.Adam(D.parameters(), lr=lr)

# Loss function
#def my_bce_loss(preds, targets):
#    return F.binary_cross_entropy(preds, targets)

#loss_fn = nn.BCEWithLogitsLoss()
loss_fn = None

if wandb_log: 
    wandb.init(project="conditional-gan-mnist")

    # Log hyperparameters
    wandb.config.update({
        "batch_size": mb_size,
        "Z_dim": Z_dim,
        "X_dim": X_dim,
        "h_dim": h_dim,
        "lr": lr,
    })

best_g_loss = float('inf')  # Initialize best generator loss
save_dir = 'checkpoints'
os.makedirs(save_dir, exist_ok=True)

#Train epochs
epochs = 10

for epoch in range(epochs):
    G, D, G_loss_avg, D_loss_avg= logisticGANTraining(G, D, loss_fn, train_loader)

    print(f'epoch{epoch}; D_loss: {D_loss_avg:.4f}; G_loss: {G_loss_avg:.4f}')

    if G_loss_avg < best_g_loss:
        best_g_loss = G_loss_avg
        torch.save(G.state_dict(), os.path.join(save_dir, 'G_best.pth'))
        torch.save(D.state_dict(), os.path.join(save_dir, 'D_best.pth'))
        print(f"Saved Best Models at epoch {epoch} | G_loss: {best_g_loss:.4f}")

    save_sample(G, epoch, mb_size, Z_dim)


# Inference    
# G.load_state_dict(torch.load('checkpoints/G_best.pth'))
# G.eval()

# save_sample(G, "best", mb_size, Z_dim)

wandb: Currently logged in as: canatilgan (canatilgan-lule-university-of-technology) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


100%|██████████| 938/938 [00:13<00:00, 71.77it/s]


epoch0; D_loss: 0.0780; G_loss: 5.6271
Saved Best Models at epoch 0 | G_loss: 5.6271


100%|██████████| 938/938 [00:12<00:00, 74.37it/s]


epoch1; D_loss: 0.0320; G_loss: 5.9000


100%|██████████| 938/938 [00:12<00:00, 76.20it/s]


epoch2; D_loss: 0.0571; G_loss: 5.4029
Saved Best Models at epoch 2 | G_loss: 5.4029


100%|██████████| 938/938 [00:09<00:00, 100.40it/s]


epoch3; D_loss: 0.0920; G_loss: 6.0309


100%|██████████| 938/938 [00:12<00:00, 74.17it/s]


epoch4; D_loss: 0.1777; G_loss: 5.4179


100%|██████████| 938/938 [00:12<00:00, 76.18it/s]


epoch5; D_loss: 0.3034; G_loss: 4.6860
Saved Best Models at epoch 5 | G_loss: 4.6860


100%|██████████| 938/938 [00:11<00:00, 79.09it/s] 


epoch6; D_loss: 0.4628; G_loss: 3.9318
Saved Best Models at epoch 6 | G_loss: 3.9318


100%|██████████| 938/938 [00:09<00:00, 100.60it/s]


epoch7; D_loss: 0.5431; G_loss: 3.6702
Saved Best Models at epoch 7 | G_loss: 3.6702


100%|██████████| 938/938 [00:12<00:00, 73.72it/s]


epoch8; D_loss: 0.6244; G_loss: 3.2221
Saved Best Models at epoch 8 | G_loss: 3.2221


100%|██████████| 938/938 [00:12<00:00, 74.48it/s]


epoch9; D_loss: 0.7293; G_loss: 2.9135
Saved Best Models at epoch 9 | G_loss: 2.9135


In [2]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import numpy as np
import tqdm
import os
import wandb


# Hyperparameters
mb_size = 64
Z_dim = 1000
h_dim = 128
lr = 1e-3

# Load MNIST data
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Lambda(lambda x: x.view(-1))  # Flatten the 28x28 image to 784
])

train_dataset = datasets.MNIST(root='../MNIST', train=True, transform=transform, download=True)
train_loader = DataLoader(train_dataset, batch_size=mb_size, shuffle=True)

X_dim = 784  # 28 x 28

# Xavier Initialization
def xavier_init(m):
    if isinstance(m, nn.Linear):
        nn.init.xavier_normal_(m.weight)
        if m.bias is not None:
            nn.init.zeros_(m.bias)

# Generator
class Generator(nn.Module):
    def __init__(self, z_dim, h_dim, x_dim):
        super(Generator, self).__init__()
        self.fc1 = nn.Linear(z_dim, h_dim)
        self.fc2 = nn.Linear(h_dim, x_dim)
        self.apply(xavier_init)

    def forward(self, z):
        h = F.relu(self.fc1(z))
        out = torch.sigmoid(self.fc2(h))
        return out

# Discriminator
class Discriminator(nn.Module):
    def __init__(self, x_dim, h_dim):
        super(Discriminator, self).__init__()
        self.fc1 = nn.Linear(x_dim, h_dim)
        self.fc2 = nn.Linear(h_dim, 1)
        self.apply(xavier_init)

    def forward(self, x):
        h = F.relu(self.fc1(x))
        out = self.fc2(h)
        return out



# Training
def logisticGANTraining(G, D, loss_fn, train_loader):
    G.train()
    D.train()

    D_loss_real_total = 0
    D_loss_fake_total = 0
    G_loss_total = 0
    t = tqdm.tqdm(train_loader)
    
    for it, (X_real, labels) in enumerate(t):
        # Prepare real data
        X_real = X_real.float().to(device)

        # Sample noise and labels
        z = torch.randn(X_real.size(0), Z_dim).to(device)
        ones_label = torch.ones(X_real.size(0), 1).to(device)
        zeros_label = torch.zeros(X_real.size(0), 1).to(device)

        # ================= Train Discriminator =================
        G_sample = G(z)
        D_real = D(X_real)
        D_fake = D(G_sample.detach())

        D_loss_real = F.softplus(-D_real).mean()  # log(1 + e^{-D(x_real)})
        D_loss_fake = F.softplus(D_fake).mean()   # log(1 + e^{D(x_fake)})
        D_loss = D_loss_real + D_loss_fake
        D_loss_real_total += D_loss_real.item()
        D_loss_fake_total += D_loss_fake.item()

        D_solver.zero_grad()
        D_loss.backward()
        D_solver.step()

        # ================= Train Generator ====================
        z = torch.randn(X_real.size(0), Z_dim).to(device)
        G_sample = G(z)
        D_fake = D(G_sample)

        G_loss = F.softplus(-D_fake).mean()  # log(1 + e^{-D(G(z))})
        G_loss_total += G_loss.item()

        G_solver.zero_grad()
        G_loss.backward()
        G_solver.step()

    # ================= Logging =================
    D_loss_real_avg = D_loss_real_total / len(train_loader)
    D_loss_fake_avg = D_loss_fake_total / len(train_loader)
    D_loss_avg = D_loss_real_avg + D_loss_fake_avg
    G_loss_avg = G_loss_total / len(train_loader)

    wandb.log({
        "D_loss_real": D_loss_real_avg,
        "D_loss_fake": D_loss_fake_avg,
        "D_loss": D_loss_avg,
        "G_loss": G_loss_avg
    })

    return G, D, G_loss_avg, D_loss_avg
    


def save_sample(G, epoch, mb_size, Z_dim):
    out_dir = "out_vanila_GAN2"
    G.eval()
    with torch.no_grad():
        z = torch.randn(mb_size, Z_dim).to(device)
        samples = G(z).detach().cpu().numpy()[:16]

    fig = plt.figure(figsize=(4, 4))
    gs = gridspec.GridSpec(4, 4)
    gs.update(wspace=0.05, hspace=0.05)

    for i, sample in enumerate(samples):
        ax = plt.subplot(gs[i])
        plt.axis('off')
        ax.set_xticklabels([])
        ax.set_yticklabels([])
        ax.set_aspect('equal')
        plt.imshow(sample.reshape(28, 28), cmap='Greys_r')

    if not os.path.exists(f'{out_dir}'):
        os.makedirs(f'{out_dir}')

    plt.savefig(f'{out_dir}/{str(epoch).zfill(3)}.png', bbox_inches='tight')
    plt.close(fig)



########################### Main #######################################
wandb_log = True
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

# Instantiate models
G = Generator(Z_dim, h_dim, X_dim).to(device)
D = Discriminator(X_dim, h_dim).to(device)

# Optimizers
G_solver = optim.Adam(G.parameters(), lr=lr)
D_solver = optim.Adam(D.parameters(), lr=lr)

# Loss function
#def my_bce_loss(preds, targets):
#    return F.binary_cross_entropy(preds, targets)

#loss_fn = nn.BCEWithLogitsLoss()
loss_fn = None

if wandb_log: 
    wandb.init(project="conditional-gan-mnist")

    # Log hyperparameters
    wandb.config.update({
        "batch_size": mb_size,
        "Z_dim": Z_dim,
        "X_dim": X_dim,
        "h_dim": h_dim,
        "lr": lr,
    })

best_g_loss = float('inf')  # Initialize best generator loss
save_dir = 'checkpoints'
os.makedirs(save_dir, exist_ok=True)

#Train epochs
epochs = 5

for epoch in range(epochs):
    G, D, G_loss_avg, D_loss_avg= logisticGANTraining(G, D, loss_fn, train_loader)

    print(f'epoch{epoch}; D_loss: {D_loss_avg:.4f}; G_loss: {G_loss_avg:.4f}')

    if G_loss_avg < best_g_loss:
        best_g_loss = G_loss_avg
        torch.save(G.state_dict(), os.path.join(save_dir, 'G_best.pth'))
        torch.save(D.state_dict(), os.path.join(save_dir, 'D_best.pth'))
        print(f"Saved Best Models at epoch {epoch} | G_loss: {best_g_loss:.4f}")

    save_sample(G, epoch, mb_size, Z_dim)


# Inference    
# G.load_state_dict(torch.load('checkpoints/G_best.pth'))
# G.eval()

# save_sample(G, "best", mb_size, Z_dim)

100%|██████████| 938/938 [00:12<00:00, 72.86it/s]


epoch0; D_loss: 0.0894; G_loss: 6.3074
Saved Best Models at epoch 0 | G_loss: 6.3074


100%|██████████| 938/938 [00:12<00:00, 76.86it/s]


epoch1; D_loss: 0.0340; G_loss: 6.5768


100%|██████████| 938/938 [00:12<00:00, 74.04it/s]


epoch2; D_loss: 0.0714; G_loss: 4.8203
Saved Best Models at epoch 2 | G_loss: 4.8203


100%|██████████| 938/938 [00:12<00:00, 73.06it/s]


epoch3; D_loss: 0.1016; G_loss: 5.4058


100%|██████████| 938/938 [00:12<00:00, 74.00it/s]


epoch4; D_loss: 0.2056; G_loss: 4.8641


In [3]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import numpy as np
import tqdm
import os
import wandb


# Hyperparameters
mb_size = 64
Z_dim = 1000
h_dim = 128
lr = 1e-3

# Load MNIST data
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Lambda(lambda x: x.view(-1))  # Flatten the 28x28 image to 784
])

train_dataset = datasets.MNIST(root='../MNIST', train=True, transform=transform, download=True)
train_loader = DataLoader(train_dataset, batch_size=mb_size, shuffle=True)

X_dim = 784  # 28 x 28

# Xavier Initialization
def xavier_init(m):
    if isinstance(m, nn.Linear):
        nn.init.xavier_normal_(m.weight)
        if m.bias is not None:
            nn.init.zeros_(m.bias)

# Generator
class Generator(nn.Module):
    def __init__(self, z_dim, h_dim, x_dim):
        super(Generator, self).__init__()
        self.fc1 = nn.Linear(z_dim, h_dim)
        self.fc2 = nn.Linear(h_dim, x_dim)
        self.apply(xavier_init)

    def forward(self, z):
        h = F.relu(self.fc1(z))
        out = torch.sigmoid(self.fc2(h))
        return out

# Discriminator
class Discriminator(nn.Module):
    def __init__(self, x_dim, h_dim):
        super(Discriminator, self).__init__()
        self.fc1 = nn.Linear(x_dim, h_dim)
        self.fc2 = nn.Linear(h_dim, 1)
        self.apply(xavier_init)

    def forward(self, x):
        h = F.relu(self.fc1(x))
        out = self.fc2(h)
        return out



# Training
def logisticGANTraining(G, D, loss_fn, train_loader):
    G.train()
    D.train()

    D_loss_real_total = 0
    D_loss_fake_total = 0
    G_loss_total = 0
    t = tqdm.tqdm(train_loader)
    
    for it, (X_real, labels) in enumerate(t):
        # Prepare real data
        X_real = X_real.float().to(device)

        # Sample noise and labels
        z = torch.randn(X_real.size(0), Z_dim).to(device)
        ones_label = torch.ones(X_real.size(0), 1).to(device)
        zeros_label = torch.zeros(X_real.size(0), 1).to(device)

        # ================= Train Discriminator =================
        G_sample = G(z)
        D_real = D(X_real)
        D_fake = D(G_sample.detach())

        D_loss_real = F.softplus(-D_real).mean()  # log(1 + e^{-D(x_real)})
        D_loss_fake = F.softplus(D_fake).mean()   # log(1 + e^{D(x_fake)})
        D_loss = D_loss_real + D_loss_fake
        D_loss_real_total += D_loss_real.item()
        D_loss_fake_total += D_loss_fake.item()

        D_solver.zero_grad()
        D_loss.backward()
        D_solver.step()

        # ================= Train Generator ====================
        z = torch.randn(X_real.size(0), Z_dim).to(device)
        G_sample = G(z)
        D_fake = D(G_sample)

        G_loss = F.softplus(-D_fake).mean()  # log(1 + e^{-D(G(z))})
        G_loss_total += G_loss.item()

        G_solver.zero_grad()
        G_loss.backward()
        G_solver.step()

    # ================= Logging =================
    D_loss_real_avg = D_loss_real_total / len(train_loader)
    D_loss_fake_avg = D_loss_fake_total / len(train_loader)
    D_loss_avg = D_loss_real_avg + D_loss_fake_avg
    G_loss_avg = G_loss_total / len(train_loader)

    wandb.log({
        "D_loss_real": D_loss_real_avg,
        "D_loss_fake": D_loss_fake_avg,
        "D_loss": D_loss_avg,
        "G_loss": G_loss_avg
    })

    return G, D, G_loss_avg, D_loss_avg
    


def save_sample(G, epoch, mb_size, Z_dim):
    out_dir = "out_vanila_GAN2"
    G.eval()
    with torch.no_grad():
        z = torch.randn(mb_size, Z_dim).to(device)
        samples = G(z).detach().cpu().numpy()[:16]

    fig = plt.figure(figsize=(4, 4))
    gs = gridspec.GridSpec(4, 4)
    gs.update(wspace=0.05, hspace=0.05)

    for i, sample in enumerate(samples):
        ax = plt.subplot(gs[i])
        plt.axis('off')
        ax.set_xticklabels([])
        ax.set_yticklabels([])
        ax.set_aspect('equal')
        plt.imshow(sample.reshape(28, 28), cmap='Greys_r')

    if not os.path.exists(f'{out_dir}'):
        os.makedirs(f'{out_dir}')

    plt.savefig(f'{out_dir}/{str(epoch).zfill(3)}.png', bbox_inches='tight')
    plt.close(fig)



########################### Main #######################################
wandb_log = True
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

# Instantiate models
G = Generator(Z_dim, h_dim, X_dim).to(device)
D = Discriminator(X_dim, h_dim).to(device)

# Optimizers
G_solver = optim.Adam(G.parameters(), lr=lr)
D_solver = optim.Adam(D.parameters(), lr=lr)

# Loss function
#def my_bce_loss(preds, targets):
#    return F.binary_cross_entropy(preds, targets)

#loss_fn = nn.BCEWithLogitsLoss()
loss_fn = None

if wandb_log: 
    wandb.init(project="conditional-gan-mnist")

    # Log hyperparameters
    wandb.config.update({
        "batch_size": mb_size,
        "Z_dim": Z_dim,
        "X_dim": X_dim,
        "h_dim": h_dim,
        "lr": lr,
    })

best_g_loss = float('inf')  # Initialize best generator loss
save_dir = 'checkpoints'
os.makedirs(save_dir, exist_ok=True)

#Train epochs
epochs = 50

for epoch in range(epochs):
    G, D, G_loss_avg, D_loss_avg= logisticGANTraining(G, D, loss_fn, train_loader)

    print(f'epoch{epoch}; D_loss: {D_loss_avg:.4f}; G_loss: {G_loss_avg:.4f}')

    if G_loss_avg < best_g_loss:
        best_g_loss = G_loss_avg
        torch.save(G.state_dict(), os.path.join(save_dir, 'G_best.pth'))
        torch.save(D.state_dict(), os.path.join(save_dir, 'D_best.pth'))
        print(f"Saved Best Models at epoch {epoch} | G_loss: {best_g_loss:.4f}")

    save_sample(G, epoch, mb_size, Z_dim)


# Inference    
# G.load_state_dict(torch.load('checkpoints/G_best.pth'))
# G.eval()

# save_sample(G, "best", mb_size, Z_dim)

100%|██████████| 938/938 [00:12<00:00, 73.52it/s]


epoch0; D_loss: 0.0710; G_loss: 6.3610
Saved Best Models at epoch 0 | G_loss: 6.3610


100%|██████████| 938/938 [00:12<00:00, 72.93it/s]


epoch1; D_loss: 0.0345; G_loss: 6.1787
Saved Best Models at epoch 1 | G_loss: 6.1787


100%|██████████| 938/938 [00:12<00:00, 74.64it/s]


epoch2; D_loss: 0.0462; G_loss: 5.3850
Saved Best Models at epoch 2 | G_loss: 5.3850


100%|██████████| 938/938 [00:12<00:00, 74.33it/s]


epoch3; D_loss: 0.0906; G_loss: 5.9882


100%|██████████| 938/938 [00:12<00:00, 75.10it/s]


epoch4; D_loss: 0.1829; G_loss: 5.0895
Saved Best Models at epoch 4 | G_loss: 5.0895


100%|██████████| 938/938 [00:12<00:00, 74.57it/s]


epoch5; D_loss: 0.3064; G_loss: 4.4196
Saved Best Models at epoch 5 | G_loss: 4.4196


100%|██████████| 938/938 [00:12<00:00, 74.04it/s]


epoch6; D_loss: 0.4397; G_loss: 3.8090
Saved Best Models at epoch 6 | G_loss: 3.8090


100%|██████████| 938/938 [00:12<00:00, 75.50it/s]


epoch7; D_loss: 0.5412; G_loss: 3.3873
Saved Best Models at epoch 7 | G_loss: 3.3873


100%|██████████| 938/938 [00:12<00:00, 74.40it/s]


epoch8; D_loss: 0.5925; G_loss: 3.1010
Saved Best Models at epoch 8 | G_loss: 3.1010


100%|██████████| 938/938 [00:12<00:00, 73.69it/s]


epoch9; D_loss: 0.6432; G_loss: 2.8428
Saved Best Models at epoch 9 | G_loss: 2.8428


100%|██████████| 938/938 [00:12<00:00, 74.90it/s]


epoch10; D_loss: 0.6640; G_loss: 2.7327
Saved Best Models at epoch 10 | G_loss: 2.7327


100%|██████████| 938/938 [00:12<00:00, 74.82it/s]


epoch11; D_loss: 0.6983; G_loss: 2.5987
Saved Best Models at epoch 11 | G_loss: 2.5987


100%|██████████| 938/938 [00:12<00:00, 74.63it/s]


epoch12; D_loss: 0.7276; G_loss: 2.4775
Saved Best Models at epoch 12 | G_loss: 2.4775


100%|██████████| 938/938 [00:12<00:00, 74.40it/s]


epoch13; D_loss: 0.7618; G_loss: 2.4057
Saved Best Models at epoch 13 | G_loss: 2.4057


100%|██████████| 938/938 [00:12<00:00, 74.16it/s]


epoch14; D_loss: 0.7581; G_loss: 2.3252
Saved Best Models at epoch 14 | G_loss: 2.3252


100%|██████████| 938/938 [00:12<00:00, 75.65it/s]


epoch15; D_loss: 0.7851; G_loss: 2.1577
Saved Best Models at epoch 15 | G_loss: 2.1577


100%|██████████| 938/938 [00:12<00:00, 76.84it/s]


epoch16; D_loss: 0.7700; G_loss: 2.1157
Saved Best Models at epoch 16 | G_loss: 2.1157


100%|██████████| 938/938 [00:10<00:00, 86.46it/s] 


epoch17; D_loss: 0.7688; G_loss: 2.1289


100%|██████████| 938/938 [00:12<00:00, 75.79it/s]


epoch18; D_loss: 0.7694; G_loss: 2.1303


100%|██████████| 938/938 [00:10<00:00, 87.08it/s] 


epoch19; D_loss: 0.7870; G_loss: 2.1314


100%|██████████| 938/938 [00:12<00:00, 74.25it/s]


epoch20; D_loss: 0.7903; G_loss: 2.0715
Saved Best Models at epoch 20 | G_loss: 2.0715


100%|██████████| 938/938 [00:12<00:00, 75.18it/s]


epoch21; D_loss: 0.8048; G_loss: 2.0142
Saved Best Models at epoch 21 | G_loss: 2.0142


100%|██████████| 938/938 [00:11<00:00, 79.85it/s] 


epoch22; D_loss: 0.8009; G_loss: 2.0593


100%|██████████| 938/938 [00:11<00:00, 84.49it/s] 


epoch23; D_loss: 0.8064; G_loss: 2.0253


100%|██████████| 938/938 [00:11<00:00, 79.65it/s] 


epoch24; D_loss: 0.7939; G_loss: 2.0240


100%|██████████| 938/938 [00:12<00:00, 75.48it/s]


epoch25; D_loss: 0.7970; G_loss: 1.9613
Saved Best Models at epoch 25 | G_loss: 1.9613


100%|██████████| 938/938 [00:12<00:00, 75.09it/s]


epoch26; D_loss: 0.7880; G_loss: 1.9552
Saved Best Models at epoch 26 | G_loss: 1.9552


100%|██████████| 938/938 [00:12<00:00, 73.65it/s]


epoch27; D_loss: 0.7858; G_loss: 2.0548


100%|██████████| 938/938 [00:12<00:00, 74.99it/s]


epoch28; D_loss: 0.7847; G_loss: 2.0497


100%|██████████| 938/938 [00:12<00:00, 74.89it/s]


epoch29; D_loss: 0.7877; G_loss: 1.9943


100%|██████████| 938/938 [00:12<00:00, 73.96it/s]


epoch30; D_loss: 0.7793; G_loss: 1.9952


100%|██████████| 938/938 [00:12<00:00, 74.08it/s]


epoch31; D_loss: 0.7757; G_loss: 2.0178


100%|██████████| 938/938 [00:12<00:00, 73.95it/s]


epoch32; D_loss: 0.7658; G_loss: 2.0256


100%|██████████| 938/938 [00:12<00:00, 75.33it/s]


epoch33; D_loss: 0.7579; G_loss: 2.0729


100%|██████████| 938/938 [00:12<00:00, 75.06it/s]


epoch34; D_loss: 0.7440; G_loss: 2.1255


100%|██████████| 938/938 [00:12<00:00, 74.31it/s]


epoch35; D_loss: 0.7376; G_loss: 2.1245


100%|██████████| 938/938 [00:12<00:00, 75.59it/s]


epoch36; D_loss: 0.7306; G_loss: 2.1425


100%|██████████| 938/938 [00:12<00:00, 75.12it/s]


epoch37; D_loss: 0.7210; G_loss: 2.1740


100%|██████████| 938/938 [00:11<00:00, 82.43it/s] 


epoch38; D_loss: 0.7085; G_loss: 2.2152


100%|██████████| 938/938 [00:12<00:00, 73.45it/s]


epoch39; D_loss: 0.7003; G_loss: 2.2271


100%|██████████| 938/938 [00:12<00:00, 77.48it/s] 


epoch40; D_loss: 0.6907; G_loss: 2.2522


100%|██████████| 938/938 [00:09<00:00, 98.30it/s] 


epoch41; D_loss: 0.6904; G_loss: 2.2826


100%|██████████| 938/938 [00:12<00:00, 74.78it/s]


epoch42; D_loss: 0.6802; G_loss: 2.3431


100%|██████████| 938/938 [00:12<00:00, 74.57it/s]


epoch43; D_loss: 0.6715; G_loss: 2.3494


100%|██████████| 938/938 [00:12<00:00, 75.74it/s]


epoch44; D_loss: 0.6676; G_loss: 2.4003


100%|██████████| 938/938 [00:12<00:00, 74.42it/s]


epoch45; D_loss: 0.6595; G_loss: 2.4485


100%|██████████| 938/938 [00:12<00:00, 74.24it/s]


epoch46; D_loss: 0.6498; G_loss: 2.4560


100%|██████████| 938/938 [00:12<00:00, 73.16it/s]


epoch47; D_loss: 0.6445; G_loss: 2.4703


100%|██████████| 938/938 [00:12<00:00, 73.66it/s]


epoch48; D_loss: 0.6430; G_loss: 2.4832


100%|██████████| 938/938 [00:10<00:00, 91.13it/s] 


epoch49; D_loss: 0.6377; G_loss: 2.5040


In comparing Binary Cross Entropy (Task 1) and Logistic Loss (Task 2) for 10 epochs, both led to stable and successful GAN training. Logistic loss showed smoother discriminator behaviour and more gradual G_loss drop, while BCE had slightly faster convergence early on. Final Generator performance was comparable, with both reaching G_loss values around 2.7–2.9 by epoch 9. Logistic loss appears more robust and theoretically stable, while BCE works well in practice and may converge faster with fewer epochs.